src/Type/Php/PregSplitDynamicReturnTypeExtension.php

```diff
@@ -36,17 +47,131 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
 
 	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
 	{
-		$flagsArg = $functionCall->getArgs()[3] ?? null;
+		$args = $functionCall->getArgs();
+		if (count($args) < 2) {
+			return null;
+		}
+		$patternArg = $args[0];
+		$subjectArg = $args[1];
+		$limitArg = $args[2] ?? null;
+		$flagArg = $args[3] ?? null;
+		$patternType = $scope->getType($patternArg->value);
+		$patternConstantTypes = $patternType->getConstantStrings();
+		$subjectType = $scope->getType($subjectArg->value);
+		$subjectConstantTypes = $subjectType->getConstantStrings();
+
+		if (
+			count($patternConstantTypes) > 0
+			&& @preg_match($patternConstantTypes[0]->getValue(), '') === false
+		) {
+			return new ErrorType();
+		}
+
+		$limits = [];
+		if ($limitArg === null) {
+			$limits = [-1];
+		} else {
+			$limitType = $scope->getType($limitArg->value);
+			foreach ($limitType->getConstantScalarValues() as $limit) {
+				if (!is_int($limit)) {
+					return new ErrorType();
+				}
+				$limits[] = $limit;
+			}
+		}
+
+		$flags = [];
+		if ($flagArg === null) {
+			$flags = [0];
+		} else {
+			$flagType = $scope->getType($flagArg->value);
+			foreach ($flagType->getConstantScalarValues() as $flag) {
+				if (!is_int($flag)) {
```

Comment:
to be consistent with limit, this might also allow numeric-string

https://3v4l.org/PqFaA